src/UnderlineNav2/UnderlineNav.tsx

```diff
@@ -138,16 +143,75 @@ export const UnderlineNav = forwardRef(
     forwardedRef
   ) => {
     const backupRef = useRef<HTMLElement>(null)
-    const newRef = (forwardedRef ?? backupRef) as MutableRefObject<HTMLElement>
+    const navRef = (forwardedRef ?? backupRef) as MutableRefObject<HTMLElement>
     const listRef = useRef<HTMLUListElement>(null)
     const moreMenuRef = useRef<HTMLDivElement>(null)
 
     const {theme} = useTheme()
 
+    function getItemsWidth(itemText: string): number {
+      return noIconChildWidthArray.find(item => item.text === itemText)?.width || 0
+    }
+
+    const swapMenuItemWithListItem = (
+      prospectiveListItem: React.ReactElement,
+      indexOfProspectiveListItem: number,
+      event: React.MouseEvent<HTMLLIElement> | React.KeyboardEvent<HTMLLIElement>,
+      callback: (props: ResponsiveProps, displayIcons: boolean) => void
+    ) => {
+      // get the selected menu item's width
+      const widthToFitIntoList = getItemsWidth(prospectiveListItem.props.children)
+      // Check if there is any empty space on the right side of the more btn
+      const availableSpace =
+        navRef.current.getBoundingClientRect().width -
+        (moreMenuRef.current?.getBoundingClientRect().width || 0) -
+        (listRef.current?.getBoundingClientRect().width || 0)
+
+      // Calculate how many items need to be pulled in to the menu to make room for the selected menu item
+      // I.e. if we need to pull 2 items in (index 0 and index 1), breakpoint (index) will return 1.
+      const index = getBreakpointForItemSwapping(widthToFitIntoList, availableSpace)
+      const indexToSliceAt = responsiveProps.items.length - 1 - index
+      // Form the new list of items
+      const itemsLeftInList = [...responsiveProps.items].slice(0, indexToSliceAt)
+      const updatedItemList = [...itemsLeftInList, prospectiveListItem]
+      // Form the new menu items
+      const itemsToAddToMenu = [...responsiveProps.items].slice(indexToSliceAt)
+      const updatedMenuItems = [...actions]
+      // Add itemsToAddToMenu array's items to the menu at the index of the prospectiveListItem and remove 1 count of items (prospectiveListItem)
+      updatedMenuItems.splice(indexOfProspectiveListItem, 1, ...itemsToAddToMenu)
+
+      // This is to prevent calling the overflowEffect function again when the responsiveProps is being updated.
+      setIsSwapping(true)
```

Comment:
## Help needed

The callback of the `swapMenuItemWithListItem` function is updating responsiveProps in the state. Because its state changes, the UnderlineNav gets re-rendered. And when it is re-rendered, it runs the `resizeObserverCallback` which runs `overflowEffect` function that is primarily how we display items in the list and in the menu with the given children as UnderlineNav.Item. 

Setting `isSwapping` state `true` here prevents the `overflowEffect` function running when the responsiveProps is being updated and we can see the items move in the screen recording which is what we need however, it also prevents running the `overflowEffect` function when it is actually needed such as re-size or selecting a visible item because there is no where `isSwapping` gets assigned to false. 

Maybe I am so absorbed in the same code for a while but I cannot seem to figure this out. I need to prevent `overflowEffect` function from running when we are swapping items and let it run in all other cases.

Reply:
Help found 💖